### PR TITLE
[#13] [Part 1] Implement the base Clean Architecture with MVVM: data and domain layer

### LINKED ISSUE #13: Implement the base Clean Architecture with MVVM: data and domain layer
## Why

The template should provide the initial base architecture to start. The selected architecture for the shared, Android and iOS modules is [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) with MVVM.

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
- #13 

## What happened 👀

- [x] Add `buildSrc` and extract dependencies
- [x] Add `Koin` DI
- [ ] Add `Ktor`

## Insight 📝

The template should provide the initial base architecture to start. The selected architecture for the shared, Android and iOS modules is [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) with MVVM.

## Proof Of Work 📹

Girl in a jacket

Screenshot 2023-08-10 at 3 48 52 PM

### CODE REVIEW

**Review by phongvhd93 — APPROVED** [review]

File: shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package co.nimblehq.kmm.template.di.module
+
+import org.koin.core.module.Module
+import org.koin.dsl.module
+
+actual fun platformModule(): Module = module {
+}
```
Comment (lydiasama) [inline shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt:8]: Can we have //TODO: comment here too. 🙏🏻

File: shared/src/androidMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package co.nimblehq.kmm.template.di.module
+
+import org.koin.core.module.Module
+import org.koin.dsl.module
+
+actual fun platformModule(): Module = module {
+
```
Comment (lydiasama) [inline shared/src/androidMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt]: Ditto here.

File: build.gradle.kts
Comment (lydiasama) [inline build.gradle.kts:1]: Could you please rebase on #16 to avoid conflict? 🙏🏻

File: build.gradle.kts [hunk 3]
```diff
@@ -1,9 +1,9 @@
 plugins {
     //trick: for the same plugin versions in all sub-modules
-    id("com.android.application").version("8.0.2").apply(false)
-    id("com.android.library").version("8.0.2").apply(false)
-    kotlin("android").version("1.8.21").apply(false)
-    kotlin("multiplatform").version("1.8.21").apply(false)
+    id(Plugins.ANDROID_APPLICATION).version(Versions.ANDROID_APPLICATION).apply(false)
+    id(Plugins.ANDROID_LIBRARY).version(Versions.ANDROID_LIBRARY).apply(false)
+    kotlin(Plugins.ANDROID).version(Versions.KOTLIN).apply(false)
+    kotlin(Plugins.MULTIPLATFORM).version(Versions.KOTLIN).apply(false)
```
Comment (lydiasama) [inline build.gradle.kts]: Optional: Use `infix` for consistency with #16 
```suggestion
    id(Plugins.ANDROID_APPLICATION) version Versions.ANDROID_APPLICATION apply(false)
    id(Plugins.ANDROID_LIBRARY) version Versions.ANDROID_LIBRARY apply(false)
    kotlin(Plugins.ANDROID) version Versions.KOTLIN apply(false)
    kotlin(Plugins.MULTIPLATFORM) version Versions.KOTLIN apply(false)
```

**Review by lydiasama — COMMENTED** [review]
Just small refactoring comments. Please rebase on #16  🙏🏻

**Comment by lydiasama** [comment]
@kaungkhantsoe Let's rebase on develop branch. 😄

**Comment by AVI5HEK** [comment]
Conflicts ⚔️

File: android/build.gradle.kts [hunk 4]
```diff
@@ -100,6 +100,12 @@ dependencies {
         implementation(TIMBER)
     }
 
+    with(Dependencies.Koin) {
+        implementation(CORE)
+        implementation(ANDROID)
+        implementation(COMPOSE)
+    }
```
Comment (luongvo) [inline android/build.gradle.kts:103]: please move this up, on top of Log

File: buildSrc/src/main/java/Configurations.kt [hunk 5]
```diff
@@ -20,7 +20,7 @@ object Plugins {
 
     const val BUILD_KONFIG = "com.codingfeline.buildkonfig"
 
-    const val COCOAPODS = "native.cocoapods"
+    const val IOS_COCOAPODS = "native.cocoapods"
```
Comment (luongvo) [inline buildSrc/src/main/java/Configurations.kt]: please reorder in alphabet

File: buildSrc/src/main/java/Dependencies.kt [hunk 6]
```diff
@@ -23,6 +23,10 @@ object Versions {
     const val KOTLIN_COROUTINES = "1.7.3"
     const val KOVER = "0.7.3"
 
+    const val KOIN = "3.3.2"
+    const val KOIN_ANDROID = "3.3.2"
+    const val KOIN_ANDROIDX_COMPOSE = "3.4.1"
```
Comment (luongvo) [inline buildSrc/src/main/java/Dependencies.kt:24]: same

File: buildSrc/src/main/java/Dependencies.kt [hunk 7]
```diff
@@ -52,4 +58,10 @@ object Dependencies {
 
         const val JUNIT = "junit:junit:${Versions.JUNIT}"
     }
+
+    object Koin {
```
Comment (luongvo) [inline buildSrc/src/main/java/Dependencies.kt]: please move this up, on top of Log

File: shared/build.gradle.kts [hunk 8]
```diff
@@ -41,14 +41,21 @@ kotlin {
     sourceSets {
         val commonMain by getting {
             dependencies {
-                //put your multiplatform dependencies here
+                // Koin
```
Comment (luongvo) [inline shared/build.gradle.kts]: this comment is unnecessary 🙏

File: shared/build.gradle.kts [hunk 9]
```diff
@@ -41,14 +41,21 @@ kotlin {
...
         val commonTest by getting {
             dependencies {
                 implementation(kotlin("test"))
             }
         }
+        val androidMain by getting {
+            dependencies {
+                implementation(Dependencies.Koin.ANDROID)
+                implementation(Dependencies.Koin.COMPOSE)
+            }
```
Comment (luongvo) [inline shared/build.gradle.kts]: I think we need them for the shared module in androidMain. We set them in the "android" module instead?

File: shared/src/androidMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package co.nimblehq.kmm.template.di.module
+
+import org.koin.core.module.Module
+import org.koin.dsl.module
+
+actual fun platformModule(): Module = module {
+    // TODO: Declare modules here
```
[the remark this answers is not in the document]
Comment (luongvo) [reply shared/src/androidMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt]: skip adding `:` https://detekt.dev/docs/rules/style/#forbiddencomment

```suggestion
    // TODO Declare modules here
```

File: android/src/main/java/co/nimblehq/kmm/template/android/di/ViewModelModule.kt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package co.nimblehq.kmm.template.android.di
+
+import org.koin.core.module.Module
+import org.koin.dsl.module
+
+val androidViewModelModule: Module
+    get() = module {
+        // TODO declare viewModel modules here
+    }
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/kmm/template/android/di/ViewModelModule.kt]:
```suggestion
val androidViewModelModule: Module = module {
    // TODO declare viewModel modules here
}
```

File: shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt (hunk 1 above)
Comment (luongvo) [reply shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt:8]: This is a valid comment 🙏

File: shared/src/commonMain/kotlin/co/nimblehq/kmm/template/di/Koin.kt [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import co.nimblehq.kmm.template.di.module.*
+import org.koin.core.KoinApplication
+import org.koin.core.context.startKoin
+import org.koin.dsl.KoinAppDeclaration
+
+fun initKoin(
+    appDeclaration: KoinAppDeclaration = {}
+): KoinApplication {
+    val dataModules = listOf(remoteModule)
+    val domainModules = listOf(repositoryModule, useCaseModule)
```
Comment (luongvo) [inline shared/src/commonMain/kotlin/co/nimblehq/kmm/template/di/Koin.kt]: repository should be in data 🙏 

```suggestion
    val dataModules = listOf(remoteModule, repositoryModule)
    val domainModules = listOf(useCaseModule)
```

**Review by luongvo — COMMENTED** [review]
The rest lgtm

File: shared/build.gradle.kts (hunk 9 above)
Comment (kaungkhantsoe) [reply shared/build.gradle.kts]: @luongvo did you mean `I don't think we need them...` ?

Comment (luongvo) [reply shared/build.gradle.kts]: @kaungkhantsoe Ah yes, my bad. `I don't think`... 🙏

File: shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt (hunk 1 above)
Comment (luongvo) [reply shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt:8]: @kaungkhantsoe did you miss this review?

**Review by luongvo — COMMENTED** [review]
2 remaining issues @kaungkhantsoe 🙏

File: shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt (hunk 1 above)
Comment (kaungkhantsoe) [reply shared/src/iosMain/kotlin/co/nimblehq/kmm/template/di/module/PlatformModule.kt:8]: Ahh I thought I updated it. Sorry. Let me update it again.

**Review by luongvo — APPROVED** [review]
All good

**Review by lydiasama — APPROVED** [review]

**Pull request merged**